### PR TITLE
Elastic Search View Class

### PR DESCRIPTION
**Pull request by kenjinp**
# NEW
`ElasticSearchView` Class
Added ES View instance query capability, as well as some fancy convenience methods. 
such as:
- filter resources in a project by types (Using AND)
- filter a resource in a project by schema 

## MISC
- Moved getResourceByID to Resource as a named import

## TODO
- `Project` method to fetch all ESViews as ESView instances

### CODE REVIEW

File: nexus-sdk-js/src/Project.ts [hunk 1]
```diff
@@ -96,16 +96,6 @@ export default class Project {
...
-      const resource = new Resource(
-        this.orgLabel,
-        this.label,
-        resourceResponse,
-      );
-      return resource;
-    } catch (e) {
-      return e;
-    }
+    return await getResource(id, this.orgLabel, this.label);
```
Comment (julienmachon) [inline nexus-sdk-js/src/Project.ts]: does `getResource` always return a resource?

File: nexus-sdk-js/src/Resource/index.ts [hunk 2]
```diff
@@ -42,6 +44,20 @@ export interface ResourceResponse extends ResourceResponseCommon {
...
+  id: string,
+  orgLabel: string,
+  projectLabel: string,
+): Promise<Resource> => {
+  try {
+    const resourceResponse: ResourceResponse = await httpGet(id, false);
+    const resource = new Resource(orgLabel, projectLabel, resourceResponse);
+    return resource;
+  } catch (e) {
+    return e;
```
Comment (julienmachon) [inline nexus-sdk-js/src/Resource/index.ts]: I would either throw an error here or return something more useful. I would also change the function signature to something like `getResource(....): Promise<Resource | ResourceError>`

File: nexus-sdk-js/src/View/ElasticSearchView/__tests__/ElasticSearchView.spec.ts [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+    );
+    const expectedQueryURL = `views/${orgLabel}/${projectLabel}/${
+      view.id
+    }/_search`;
+    expect(view.queryURL).toEqual(expectedQueryURL);
+  });
+
+  describe('query()', () => {
+    const mockHttpPost = <jest.Mock<typeof httpPost>>httpPost;
+    const mockHttpGet = <jest.Mock<typeof httpGet>>httpGet;
```
Comment (julienmachon) [inline nexus-sdk-js/src/View/ElasticSearchView/__tests__/ElasticSearchView.spec.ts]: :kissing_heart:

File: nexus-sdk-js/src/View/ElasticSearchView/index.ts [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+   * @memberof ElasticSeachView
+   */
+  async query(
+    elasticSearchQuery: Object,
+    pagination?: PaginationSettings,
+  ): Promise<PaginatedList<Resource>> {
+    const requestURL = pagination
+      ? `${this.queryURL}?from=${pagination.from}&size=${pagination.size}`
+      : this.queryURL;
+    const response: ElasticSearchViewQueryResponse = await httpPost(
```
Comment (julienmachon) [inline nexus-sdk-js/src/View/ElasticSearchView/index.ts]: Remember than `httpPost` can crash, so you should wrap this in a `try/catch` statement. We need to change the signature of our `http` method and make them return `Promise<T | Error>` so typescript can catch stuff like that.

File: nexus-sdk-js/src/Project.ts (hunk 1 above)
Comment (kenjinp) [reply nexus-sdk-js/src/Project.ts]: It does, or throws an error

File: nexus-sdk-js/src/View/ElasticSearchView/__tests__/ElasticSearchView.spec.ts (hunk 3 above)
Comment (kenjinp) [reply nexus-sdk-js/src/View/ElasticSearchView/__tests__/ElasticSearchView.spec.ts]: 👨‍🍳

File: nexus-sdk-js/src/Project/index.ts [hunk 5]
```diff
@@ -96,16 +100,29 @@ export default class Project {
   }
 
   async getResource(id: string): Promise<Resource> {
+    return await getResource(id, this.orgLabel, this.label);
+  }
+
+  // The current API does not support pagination / filtering of views
+  // This should be fixed when possible and converted to signituare
+  // Promise<PaginatedList<ElasticSearchView>>
+  async getElasticSearchViews(): Promise<ElasticSearchView[]> {
```
Comment (julienmachon) [inline nexus-sdk-js/src/Project/index.ts]: I would rename this to `listElasticSearchViews`.

Comment (kenjinp) [reply nexus-sdk-js/src/Project/index.ts]: you got it dude!

Comment (julienmachon) [reply nexus-sdk-js/src/Project/index.ts]: also, can you add a `getElasticSearchView(id?: string = 'document')` method that returns the specific Elastic view (you need to manually check if it is an `ElasticView` in the types) or the default `ElasticView` if no ID is specified.

I will reflect that with `SparqlViews` methods.

Comment (julienmachon) [reply nexus-sdk-js/src/Project/index.ts]: OK, looks like the api endpoint hasn't been implemented yet, so let's do this later. I've created an issue for it BlueBrain/nexus#333

**Pull request merged**